### PR TITLE
add basic functionality to application

### PR DESCRIPTION
**Pull request by aleks-and**

### CODE REVIEW

File: src/js/components/LogItem/LogItem.tsx [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    LOG_TYPE_ERROR, PARAMETERS_TAB_VIEW_TYPE_JSON, PARAMETERS_TAB_VIEW_TYPE_TREE
+} from '../../constants';
+
+import {getSubLogColor, getLogColor, createSubLogTitle, createLogTitle, createCommandHeaderDocLink} from './utils';
+
+import {JSONTreeTheme} from './constants';
+
+
+const LogItem = ({
+    log,
```
Comment (DeadByElpy) [inline src/js/components/LogItem/LogItem.tsx:31]: need to discuss

**Comment by DarkPark** [comment]
wrong title in index.html

File: src/js/components/LogItem/utils/createCommandHeaderDocLink.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const baseUrl = 'https://github.com/jooby-dev/jooby-docs/blob/main/docs/message.md#';
+
+
+export const createCommandHeaderDocLink = (headerLength: number) => {
```
Comment (DarkPark) [inline src/js/components/LogItem/utils/createCommandHeaderDocLink.tsx:4]: let's discuss default export here and in other places

File: src/js/components/LogItem/utils/decimalToHex.tsx [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export const decimalToHex = (decimal: number, prefix = '0x') => (
```
Comment (DarkPark) [inline src/js/components/LogItem/utils/decimalToHex.tsx:1]: how about use it from jooby-codec/utils?

File: src/js/components/CommandParameterEditor/utils/isValidJson.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+export const isValidJson = (jsonString: string): boolean => {
+    try {
+        const parsedObject = JSON.parse(jsonString);
+
+        return parsedObject && typeof parsedObject === 'object';
```
Comment (DarkPark) [inline src/js/components/CommandParameterEditor/utils/isValidJson.tsx:5]: if parsedObject is nullish then isValidJson does not return boolean

File: src/js/components/CommandPanel/CommandPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,723 @@
...
+
+        try {
+            data = message.fromHex(buffer, Number(selectedParseButtonIndex), getHardwareType(hardwareType));
+        } catch (error) {
+            parseError = error;
+        }
+
+        if (data) {
+            data.commands = data.commands.map((commandData) => {
+                return {
```
Comment (DarkPark) [inline src/js/components/CommandPanel/CommandPanel.tsx:255]:
```suggestion
```
?

File: src/js/components/CommandPanel/CommandPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,723 @@
...
+            errorMessage: parseError?.message,
+            type: parseError ? LOG_TYPE_ERROR : LOG_TYPE_MESSAGE,
+            id: uuidv4()
+        };
+
+        setLogs(prevLogs => [log, ...prevLogs]);
+    };
+
+    const onPreparedCommandDragEnd = (result) => {
+        if (!result.destination) return;
```
Comment (DarkPark) [inline src/js/components/CommandPanel/CommandPanel.tsx:293]: better in 3 lines

File: src/js/components/CommandPanel/CommandPanel.tsx [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+            setTimeout(() => setRecentlyEditedCommandId(null), 2000);
+        }
+    };
+
+    const handleCancelEditingCommandClick = () => {
+        setEditingCommandId(null);
+        setParameters('');
+    };
+
+
```
Comment (DarkPark) [inline src/js/components/CommandPanel/CommandPanel.tsx:339]:
```suggestion
```

eslint should detect this

File: src/js/components/CommandPanel/CommandPanel.tsx [hunk 8]
```diff
@@ -0,0 +1,723 @@
...
+                                                p: 0,
+                                                borderWidth: '1px',
+                                                borderStyle: 'solid',
+                                                borderColor: 'divider'
+                                            }}
+                                        >
+                                            {preparedCommands.map((preparedCommand, index) => (
+                                                <Draggable key={preparedCommand.id} draggableId={preparedCommand.id} index={index}>
+                                                    {(provided) => (
+                                                        <ListItem
```
Comment (DarkPark) [inline src/js/components/CommandPanel/CommandPanel.tsx:589]: isn't it too nested?

File: src/js/utils/createCommandDirectionIcon.tsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import {ArrowDownward as ArrowDownwardIcon, QuestionMark as QuestionMarkIcon} from '@mui/icons-material';
+import {yellow} from '@mui/material/colors';
+import {DIRECTION_TYPE_DOWNLINK, DIRECTION_TYPE_UPLINK} from '../constants';
+
+
+export const createCommandDirectionIcon = (directionType: typeof DIRECTION_TYPE_DOWNLINK | typeof DIRECTION_TYPE_UPLINK | undefined) => {
+    switch (directionType) {
+        case undefined:
```
[the remark this answers is not in the document]
Comment (DeadByElpy) [reply src/js/utils/createCommandDirectionIcon.tsx:8]: rework when jooby-codec will export Command & UnknownCommand

File: src/js/components/CommandPanel/utils/createCtrlEnterSubmitHandler.tsx [hunk 10]
```diff
@@ -0,0 +1,11 @@
+export const createCtrlEnterSubmitHandler = (onSubmit: () => void) => (
```
[the remark this answers is not in the document]
Comment (DeadByElpy) [reply src/js/components/CommandPanel/utils/createCtrlEnterSubmitHandler.tsx:1]: add todo

File: .eslintrc.cjs [hunk 11]
```diff
@@ -53,5 +53,14 @@ module.exports = {
         '@typescript-eslint/object-curly-spacing': ['error', 'never'],
         '@typescript-eslint/comma-dangle': ['error', 'never'],
         '@typescript-eslint/space-before-function-paren': ['error', {anonymous: 'always', named: 'always'}],
+
+        "jsx-quotes": ["error", "prefer-double"],
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply .eslintrc.cjs:57]: fixed

File: readme.md [hunk 12]
```diff
@@ -1,2 +1,6 @@
 Jooby message encoders/decoders GUI
 ===================================
+
+[![GitHub Pages](https://img.shields.io/badge/Live%20Demo-View-green)](https://jooby-dev.github.io/jooby-gui/)
+
+This application is the graphical interface of the [jooby-codec](https://github.com/jooby-dev/jooby-codec) library.
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply readme.md:6]: applied

**Comment by aleks-and** [comment]
> wrong title in index.html

fixed

File: .github/workflows/deploy.yml [hunk 13]
```diff
@@ -32,7 +29,6 @@ jobs:
 
           - run: npm ci
           - run: npx vite build
-          - run: touch ./dist/.nojekyll
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply .github/workflows/deploy.yml:35]: restored

File: src/js/utils/createCommandDirectionIcon.tsx [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+            return <QuestionMarkIcon sx={{mr: 2, color: 'grey.700'}} />;
+
+        case DIRECTION_TYPE_UPLINK:
+            return <ArrowDownwardIcon sx={{mr: 2, transform: 'rotate(180deg)', color: `${yellow[800]}`}} />;
+
+        case DIRECTION_TYPE_DOWNLINK:
+            return <ArrowDownwardIcon color="success" sx={{mr: 2}} />;
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/utils/createCommandDirectionIcon.tsx:15]: applied

File: src/js/utils/index.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+export {createCommandDocLink} from './createCommandDocLink';
+
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/utils/index.ts:2]: no longer relevant

File: src/js/utils/createCommandDocLink.tsx [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import {Command} from '../types';
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/utils/createCommandDocLink.tsx:1]: fixed

File: src/js/contexts/SnackbarContext.tsx [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+        }, 200);
+    };
+
+    useEffect(() => {
+        return () => {
+            if (timeoutId.current) {
+                clearTimeout(timeoutId.current);
+            }
+        };
+    }, []);
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/contexts/SnackbarContext.tsx:73]: applied

File: src/js/contexts/SnackbarContext.tsx [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    const [snackbarMessage, setSnackbarMessage] = useState('');
+    const [snackbarSeverity, setSnackbarSeverity] = useState<Severity>(SEVERITY_TYPE_SUCCESS);
+    const [snackbarDuration, setSnackbarDuration] = useState(DEFAULT_SNACKBAR_DURATION);
+    const snackbarQueue = useRef<Array<{message: string; duration: number, severity: Severity}>>([]);
+    const timeoutId = useRef<number | null>(null);
+
+    const processQueue = () => {
+        if (snackbarQueue.current.length > 0) {
+            const nextSnackbar = snackbarQueue.current.shift();
+            if (nextSnackbar) {
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/contexts/SnackbarContext.tsx:33]: applied

File: src/js/components/LogItem/utils/createCommandHeaderDocLink.tsx (hunk 2 above)
Comment (aleks-and) [reply src/js/components/LogItem/utils/createCommandHeaderDocLink.tsx:4]: discussed

changed to export by default

File: src/js/components/App.tsx [hunk 19]
```diff
@@ -1,4 +1,22 @@
...
+
+import {Log} from '../types';
+
+
+const App = () => {
+    const [logs, setLogs] = useState<Array<Log>>([]);
+
+    return (
+        <Box sx={{display: 'flex', flexGrow: 1, '& > *': {minWidth: 0}}}>
+            <CommandPanel setLogs={setLogs} />
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/App.tsx:15]: discussed

left it as it is for now

File: src/js/components/HighlightedText.tsx [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+import {Box} from '@mui/material';
+
+
+const HighlightedText = ({
+    children,
+    color = 'inherit',
+    fontSize = 'inherit',
+    fontWeight = 'fontWeightMedium',
+    isMonospacedFont = false
+}: {
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/HighlightedText.tsx:11]: done

File: src/js/components/CommandPanel/CommandPanel.tsx (hunk 5 above)
Comment (aleks-and) [reply src/js/components/CommandPanel/CommandPanel.tsx:257]: applied

File: src/js/components/LogItem/utils/decimalToHex.tsx (hunk 3 above)
Comment (aleks-and) [reply src/js/components/LogItem/utils/decimalToHex.tsx:1]: discussed

for now, we decided to leave it as it is

File: src/js/types/index.ts [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+export type HandleParametersTabChange = (event: React.SyntheticEvent, newValue: ParametersTab) => void;
+
+export interface ShowSnackbarParams {
+    message: string;
+    duration?: number;
+    severity?: Severity;
+}
+
+export interface Log {
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/types/index.ts:48]: renamed to `LogItem`

File: src/js/constants/index.ts [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+export const parseButtonNameMap = {
+    [AUTO]: 'Parse (auto)',
+    [DOWNLINK]: `Parse (${directions[DOWNLINK]})`,
+    [UPLINK]: `Parse (${directions[UPLINK]})`
+} as const;
+
+export const hardwareTypeList = Object.entries(constants.hardwareTypes).map(([key, value]) => ({
+    label: key,
+    value
+}));
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/constants/index.ts:43]: applied

File: src/js/main.tsx [hunk 23]
```diff
@@ -1,11 +1,49 @@
 import React from 'react';
 import {createRoot} from 'react-dom/client';
+import CssBaseline from '@mui/material/CssBaseline';
+import {createTheme, ThemeProvider} from '@mui/material/styles';
+
+import {SnackbarProvider} from './contexts/SnackbarContext';
 import App from './components/App';
 import '../sass/main.scss';
 
+declare module '@mui/material/styles' {
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/main.tsx:10]: `TypeBackground` - interface, as used by Material UI. I need to extend this interface to be able to add new properties to the theme (background.filled, background.filledHover).

File: src/js/main.tsx [hunk 24]
```diff
@@ -1,11 +1,49 @@
...
+    interface TypeBackground {
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/main.tsx:11]: https://github.com/jooby-dev/jooby-gui/pull/1#discussion_r1190282255

File: src/js/main.tsx [hunk 25]
```diff
@@ -1,11 +1,49 @@
...
+        }
+    }
+});
+
 
 createRoot(document.getElementById('root') as HTMLElement).render(
-    <React.StrictMode>
-        <App />
-    </React.StrictMode>
+    // <React.StrictMode>
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/main.tsx:41]: done

File: src/js/components/App.tsx [hunk 26]
```diff
@@ -1,4 +1,22 @@
-const App = () => <div className="App"></div>;
+import {useState} from 'react';
+import {Box} from '@mui/material';
+
+import LogsPanel from './LogsPanel';
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/App.tsx:4]: fixed

File: readme.md [hunk 27]
```diff
@@ -1,2 +1,6 @@
 Jooby message encoders/decoders GUI
 ===================================
+
+[![GitHub Pages](https://img.shields.io/badge/Live%20Demo-View-green)](https://jooby-dev.github.io/jooby-gui/)
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply readme.md:4]: applied

File: src/js/components/CommandPanel/constants.ts [hunk 28]
```diff
@@ -0,0 +1,11 @@
+import {commands} from 'jooby-codec';
+import {directions} from '../../constants';
+
+
+export const preparedCommandList = [...Object.values(commands.uplink), ...Object.values(commands.downlink)]
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/CommandPanel/constants.ts:5]: done

File: src/js/constants/index.ts [hunk 29]
```diff
@@ -0,0 +1,43 @@
+import {constants} from 'jooby-codec';
+
+
+const {AUTO, DOWNLINK, UPLINK} = constants.directions;
+
+
+export const DIRECTION_TYPE_AUTO = AUTO;
+
+export const DIRECTION_TYPE_DOWNLINK = DOWNLINK;
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/constants/index.ts:9]: discussed

unnecessary constants removed

File: src/js/utils/createCommandDirectionIcon.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/utils/createCommandDirectionIcon.tsx:8]: ok

File: src/js/components/CommandPanel/types.ts [hunk 30]
```diff
@@ -0,0 +1,51 @@
+import {DIRECTION_TYPE_AUTO, DIRECTION_TYPE_DOWNLINK, DIRECTION_TYPE_UPLINK} from '../../constants';
+
+
+export type HardwareType = object | null;
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/CommandPanel/types.ts:4]: to do added

File: src/js/components/CommandPanel/types.ts [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+
+export type CommandList = Array<object>;
+
+export type CommandExampleList = Array<object>;
+
+export type PreparedCommands = Array<object>;
+
+export type AnchorButtonGroupList = HTMLDivElement | null;
+
+export type SelectedParseButtonIndex =
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/CommandPanel/types.ts:16]: discussed

File: src/js/components/CommandPanel/utils/createCtrlEnterSubmitHandler.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/CommandPanel/utils/createCtrlEnterSubmitHandler.tsx:1]: to do added

File: src/js/components/CommandPanel/CommandPanel.tsx (hunk 8 above)
Comment (aleks-and) [reply src/js/components/CommandPanel/CommandPanel.tsx:591]: Yes, the `CommandPanel` component needs to be divided into smaller components.

Discussed, will be implemented in the future.

File: src/js/components/CommandPanel/types.ts [hunk 32]
```diff
@@ -0,0 +1,51 @@
+import {DIRECTION_TYPE_AUTO, DIRECTION_TYPE_DOWNLINK, DIRECTION_TYPE_UPLINK} from '../../constants';
+
+
+export type HardwareType = object | null;
+
+export type CommandExample = object | null;
+
+export type CommandList = Array<object>;
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/CommandPanel/types.ts]: Discussed. Unnecessary interfaces and types have been removed. The remaining ones are renamed according to our team's convention.

File: src/js/components/CommandPanel/types.ts [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+export type PreparedCommands = Array<object>;
+
+export type AnchorButtonGroupList = HTMLDivElement | null;
+
+export type SelectedParseButtonIndex =
+    | typeof DIRECTION_TYPE_AUTO
+    | typeof DIRECTION_TYPE_DOWNLINK
+    | typeof DIRECTION_TYPE_UPLINK;
+
+export type HandleHardwareTypeChange = (
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/components/CommandPanel/types.ts]: https://github.com/jooby-dev/jooby-gui/pull/1#discussion_r1191730098

File: src/js/types/index.ts [hunk 34]
```diff
@@ -0,0 +1,56 @@
+import {
+    SEVERITY_TYPE_SUCCESS, SEVERITY_TYPE_WARNING, SEVERITY_TYPE_ERROR, LOG_TYPE_ERROR, LOG_TYPE_MESSAGE,
+    PARAMETERS_TAB_VIEW_TYPE_TREE, PARAMETERS_TAB_VIEW_TYPE_JSON
+} from '../constants';
+
+type LogType = typeof LOG_TYPE_ERROR | typeof LOG_TYPE_MESSAGE;
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/types/index.ts]: done

File: src/js/types/index.ts [hunk 35]
```diff
@@ -0,0 +1,56 @@
...
+
+export type HandleLogsLimitExceededDialogClose = () => void;
+
+export type HandleLogClick = (id: string) => void;
+
+export type HandleDeleteLogClick = (event: React.SyntheticEvent, id: string) => void;
+
+export type HandleParametersTabChange = (event: React.SyntheticEvent, newValue: ParametersTab) => void;
+
+export interface ShowSnackbarParams {
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/types/index.ts]: done

File: src/js/types/index.ts [hunk 36]
```diff
@@ -0,0 +1,56 @@
...
+
+export type ParametersTab = typeof PARAMETERS_TAB_VIEW_TYPE_TREE | typeof PARAMETERS_TAB_VIEW_TYPE_JSON;
+
+export type SetLogs = (logs: Array<Log>) => void;
+
+export type Logs = Array<Log>;
+
+export type Command = object | null;
+
+export type LogCommand = object;
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/types/index.ts]: https://github.com/jooby-dev/jooby-gui/pull/1#discussion_r1191730098

File: src/js/contexts/SnackbarContext.tsx [hunk 37]
```diff
@@ -0,0 +1,95 @@
...
+import {Snackbar, Alert, SnackbarCloseReason} from '@mui/material';
+
+import {ShowSnackbarParams, Severity} from '../types';
+import {SEVERITY_TYPE_SUCCESS} from '../constants';
+
+
+const DEFAULT_SNACKBAR_DURATION = 3000;
+
+
+interface SnackbarContextData {
```
[the remark this answers is not in the document]
Comment (aleks-and) [reply src/js/contexts/SnackbarContext.tsx]: done

File: .eslintrc.cjs [hunk 38]
```diff
@@ -53,5 +53,15 @@ module.exports = {
         '@typescript-eslint/object-curly-spacing': ['error', 'never'],
         '@typescript-eslint/comma-dangle': ['error', 'never'],
         '@typescript-eslint/space-before-function-paren': ['error', {anonymous: 'always', named: 'always'}],
+
+        'no-multiple-empty-lines': ['error', {max: 2, maxEOF: 0, maxBOF: 0}],
```
Comment (DeadByElpy) [inline .eslintrc.cjs:57]: why maxEOF = 0?

**Review by DeadByElpy — APPROVED** [review]

**Pull request merged**